### DISCUSSION TITLE
hono vs nitro / Javascript runtime

### DISCUSSION THREAD

**Discussion by aheissenberger** [wakujs/waku#522] [General]
What have been the reasons to use hono and not [nitro](https://github.com/unjs/nitro)?

Other frameworks (e.g. [Angular / Analog](https://analogjs.org), [solid start / vinxi](https://start.solidjs.com/getting-started/what-is-solidstart)) use nitro which helped them to reduce the effort to implement  deployment adapters for each target platform and offers other features as a file system router, unstorage,... - see https://unjs.io

**Comment by dai-shi** [comment]
It's originally suggested by someone.
I don't investigate how nitro works, but do you think it will help a lot for Waku?
hono has adapters, but they are for edge functions, and we need serverless functions (in terms of Vercel). And, we support Node.js too.
We don't need a framework (which is what I see nitro), like Waku will have a filesystem router. It doesn't feel like nitro deployment adapters work for us like a magic. Correct me if I'm wrong.

In short, we want something minimal.

**Reply by aheissenberger** [reply]
Have a look at how [vinxi](https://github.com/nksaraf/vinxi) is using nitro ,maybe you can find some solutions in this code base to fix the problems with bundling existing frontend libraries with CommonJS parts. Solid Start switched from their own implementation to vinxi/nitro and have been able to cut down from thousand lines of code to a few line of code.

Nitro is not a framework,it is a collection of libraries used by nuxtjs - similar to Hono you can decide which of the tools (e.g. h3) you need.

**Reply by dai-shi** [reply]
I can look a closer look later, but are you really suggesting to move from hono to nitro?
It doesn't feel like it reduces too much effort, as our philosophy is to develop a minimal framework with RSC (the core doesn't include a router, at the moment). In any case, we need our own logic for our api, and it doesn't change much between hono and nitro. If nitro provides smaller capability than hono, i might be interested, but it seems the other way around.

> Solid Start switched from their own implementation to vinxi/nitro and have been able to cut down from thousand lines of code to a few line of code.

Though, that sounds convincing.

**Reply by Aslemammad** [reply]
I love Vinxi, but using it maybe reduce from the "minimalist" mission Waku has, since Vinxi has so much that we do not need since it's a full featured library. 

- Hono is the most minimalistic one, but it's not vite friendly and Nitro's experience is much better (and it has server HMR as far as I know) 
- Nitro might make the most sense now since it reduces boilerplate and also supports so many things that we do not have to implement (endless adapters). It also makes Waku run in Bun and Deno too.

One downside to Nitro/Vinxi is that pnpm start would be vite too if I guess properly, but when I discussed that before with Daishi, one of the primary goals was to keep pnpm start (waku start) agnostic rather than relying on vite for that part.

After few bug fixes now I'm working on, I might try migrating over to Nitro since I believe it makes the most sense, but anyway, let me know folks.

**Reply by dai-shi** [reply]
Thanks for sharing your thoughts.

> endless adapters

Yeah, that's probably most annoying for me. But, at the same time, if we add our own rules, we'd need to do it anyways.

> relying on vite for that part

Yes, I want to hide vite & hono internally, and still allow to make vite plugins and hono middlewares usable if users want.

> After few bug fixes now I'm working on, I might try migrating over to Nitro

Yeah, seeing it in action might be convincing. You are welcome to try it once we settle on some stability.
However...

> using it maybe reduce from the "minimalist" mission Waku has

...to me, this is important. We don't prefer big dependencies (or even small ones). So, that's a big minus.

**Reply by aheissenberger** [reply]
@dai-shi I do not suggest to include vinxi, but it has some great abstractions which could help WAKU too. That's why I thought it would help if you take some time to study the source code :-)

**Reply by dai-shi** [reply]
Do you suggest to move from hono to nitro?
(I mean I trust you in the sense that you read Waku source code.)

**Reply by aheissenberger** [reply]
I am not an expert on hono or unjs/h3.
I appreciate @Aslemammad doing a prototype, as in my opinion this is the only way to get the insights for such a decision.

Based on the history of both libraries and their current goals, unjs/h3 fits your "minimal" requirements  as all libraries in the unjs pool are single purpose and go even further by allowing to remove parts of the code thru tree shaking. Hono is currently moving into many fields with their own JSX implementation and HonoX.

Based on your roadmap, a change in this area will have a strong impact on most of the tasks (e.g. middleware, accessing request context in server components) grouped under v0.20.0.

When it comes to priorities I personally think that fixing the root causes for bundling problems with existing react libraries should be prio one, as this will be the nightmare for most people trying to use WAKU for their projects. The goal here should be that libraries that work with Next.js app folder should also work with WAKU without special handling in the vite config.

**Reply by dai-shi** [reply]
That's why I'm asking. It's a good chance to tackle before v0.20.0.
The current hono dependency is pretty minimal.

> Hono is currently moving into many fields 

which is actually a plus, if we can use their ecosystem.
In general, the question is what's post expressjs.
But I agree Hono is going too broadly far more than what we need.

> When it comes to priorities I personally think that fixing the root causes for bundling problems with existing react libraries should be prio one

That's true and @sandren says so too. But, it has nothing to do with web server implementation.

> unjs/h3 fits your "minimal" requirements

h3 looks good.
👀

**Reply by dai-shi** [reply]
> (endless adapters)

That's a big motivation of mine, but it doesn't seem to help much.
https://h3.unjs.io/adapters

We would still need wrappers on our end, and our primary adapters for Vercel serverless and Netlify functions aren't supported.

**Reply by dai-shi** [reply]
now reading the source code.

**Reply by dai-shi** [reply]
btw, any other web server alternatives to h3 and hono, which support web standard + node.js adapter?

**Reply by dai-shi** [reply]
h3 doesn't look bad, and hono doesn't look bad either.
both provide more capabilities than I want and I almost feel like I'd like to write my own.

here's my conclusion:
- i don't decide which one to go in the long term, yet.
- i will keep using hono for v0.20.0.
- IMPORTANT: but, I will keep hiding hono for user-facing APIs.
- maybe I'd go with koa-like middleware API.

Thanks for the suggestions and discussions. They gave me an insight for v0.20.0 apis.

**Reply by dai-shi** [reply]
some latest thoughts: as our user-facing apis are independent from hono, we can replace hono with something else. maybe we should revisit this before reaching v1-alpha. one of the concerns is how to support various cloud services. I would like to introduce plugin-like system for adapters instead of including all in the core repo. it will be a good timing to replace hono when we do that.

**Reply by jkhaui** [reply]
Regarding Vite support and adapters, it appears Hono already supports this with modular Vite plugins? https://github.com/honojs/vite-plugins/tree/main/packages/dev-server

I'll do some research later. Hopefully I can also provide some insight. 

I don't know much about Vinxi/Nitro, but my initial feeling is **maybe** resources are better allocated to sticking with Hono (can also refactor to use Hono's Vite dev server plugin & adapter system)

**Reply by dai-shi** [reply]
Just a note about the current design policy.
Originally I thought about tight integration with Hono and Vite, but I feel like they should be less coupled. Our hono integration is very thin, and even for Vite, we structure our code so that only some parts depend on Vite.

**Reply by jkhaui** [reply]
> Just a note about the current design policy.
> 
> Originally I thought about tight integration with Hono and Vite, but I feel like they should be less coupled. Our hono integration is very thin, and even for Vite, we structure our code so that only some parts depend on Vite.

Hmm, if I'm understanding correctly, may I know the rationale for wanting minimal coupling to Vite?

I think Waku has big potential appeal to users wanting the latest react and RSC features but outside the next + webpack/turbopack ecosystem. These users will likely have had some prior exposure to Vite

However, the main trade-off I see is in this scenario is: coupling is beneficial if it improves overall cohesion of the framework.

What I mean is, by embracing some of Vite's opinionated conventions, all efforts with Waku can be devoted to developing the main featureset. With Vite providing a battle-tested foundation & guardrails, Waku can then develop its own adapter and integrations/plugin layer on top of Vite. And developers can also bring in "vanilla" Vite plugins they're used to working with from previous projects. But the framework is still minimal (IMO).

Astro executes this design really well. End-users usually never need to know they're using Vite, but power users are free to tap into familiar Vite features if they choose

**Reply by dai-shi** [reply]
I don't have any plan to avoid Vite, but the rationale is RSC isn't tied to Vite. And, our target includes non-Nodejs env, so production build is independent from Vite. It makes sense to me that the structure is to separate to Vite dependent part and independent part.

I guess Astro design should be close and there must be something I should learn.

> End-users usually never need to know they're using Vite, but power users are free to tap into familiar Vite features if they choose

I like the idea pretty much.

**Reply by aretrace** [reply]
Stick with hono 😁. While I'm sure framework authors appreciate the ready made vinxi SDK, I already see a potential monoculture developing in the js (meta)framework ecosystem revolving around unjs/h3 due to vinxi. One of the big reasons some js devs use other (meta)frameworks is because of the dissatisfaction they have with a default (and usually only) http server implementation of some [particular (meta)framework](https://gprivate.com/6ctbm). Waku has a opportunity to really stand out here, and I believe hono is an excellent choice when looking for an express-like alternative. Hono has a more focused developer base (h3 is one of many unjs projects), is more [popular](https://npmtrends.com/h3-vs-hono) (admittedly not the best deciding factor), and in my opinion is a better http framework (comparing their performance, api design, and availability of deployment targets).

**Reply by dai-shi** [reply]
Good to hear that.

**Comment by dai-shi** [comment]
https://github.com/nksaraf/vinxi
I'm having hard time reading the repo. Maybe, it's built with a different mental model from mine.

https://github.com/nksaraf/vinxi/tree/main/examples/react/rsc/spa
The example helped a bit. It explains, something like what you said, providing parts and users can build using all or some of them.
We plan to do something with #219 and further, which might be seeking similar use cases.

I'll dig it further when I get time.
